### ISSUE TITLE
Packing for HF model

### ISSUE THREAD

**Issue by saareliad** [google-research/text-to-text-transfer-transformer#365]
Hi, I see that for the HF model there is no data-packing albite in the experiments you did use packing.
https://github.com/google-research/text-to-text-transfer-transformer/blob/5053a463eac3423284c327bf36e61988189239c1/t5/models/hf_model.py#L114

I want to use packing as you did, however, I noticed that after packing, there are several extra inputs created which seems like mesh_tensorflow internals:
https://github.com/tensorflow/mesh/blob/6a812c8bb847e081e976533ed497c7c5016bb1ec/mesh_tensorflow/transformer/dataset.py#L474-L504

Is such packing is possible for the HF model or there are some inherent limitations?

**Comment by craffel** [comment]
> I did not see a single HF example of training with T5 which achieved comparable sota results to what you present here with mesh

I was able to match the results for the Base model between HF and MTF. You can't train the 11B model in HF so of course you won't be able to get SoTA with HF.

> z_loss which I implemented on HF model

This should have a negligible impact on performance - it is mainly for numerical instability.

> adafactor - took from another source

Note that HF removed the adafactor accumulator variables from the checkpoints, whereas in MTF we load them in from the pre-trained model when fine-tuning. This hasn't seemed to make a big difference though.

> packing [WIP]

Packing makes no difference at all apart from efficiency. Assuming you do an update on a packed batch with N examples and a padded (unpacked) batch with N examples, the resulting gradient step will be identical.

**Comment by saareliad** [comment]
@craffel thanks for the replay 👍 ,

We actually can (and currently running) the large models (first 3B then 11B) in fp32 + beta1>0 with our own memory-efficient algorithm so I want to get close-enough results to empirically prove a point (for an academic paper). I guess we only want close-enough results (If it would be too far from T5-paper no one would believe us...) but not necessarily the exact same. I also run a baseline with "standard" training and the same seed, hyperparameters, and so on.

### HF accuracy:
Only yesterday I managed to restore (and actually improve a little) the reported RTE result for T5-small on the HF-original model with our algorithm and your data-pipeline and given you say the base model also worked for you that is certainly good news. (Can I ask for which task/seq_len/batch you made T5-base-HF work with?)

The collab examples on HF actually fine-tune squad to lower result than the C4 checkpoint.
I tried doing a similar thing with T5-small and T5-large with batch sizes 256 and 128 and seq_len=(384,32) and ~4000 steps and results were not close to yours. For T5-large they actually got worse than the C4 checkpoint.


### Packing:
I think _in practice_ it can make a difference, let me explain:
the problem is we don't know the N for each packed batch and N changes per batch. For example, if N=331 for a given packed batch and batch_size=128, without packing it is quite a headache to replicate, (although I agree is theoretically possible).

### Loading Adafactor:
I looked at the MTF code but didn't notice where the loading of the Adafactor parameters is but I'll keep looking. I actually never saw someone release their pre-trained optimizer state before (for adam the optimizer state is 2-3 times bigger than the model, but I saw you used beta1=0 in Adafactor in all experiments so it should be very small actually).

If you can somehow quantify the generalization difference you saw (even raw estimation) that would be useful so we could know what to expect.

**Comment by craffel** [comment]
> (Can I ask for which task/seq_len/batch you made T5-base-HF work with?)

I tried for a few GLUE tasks. The CoLA fine-tuning recipe in the hf_model.py docstring should give comparable results to MTF fine-tuning. https://github.com/google-research/text-to-text-transfer-transformer/blob/master/t5/models/hf_model.py#L16

> The collab examples on HF actually fine-tune squad to lower result than the C4 checkpoint.

I'm not sure what colab you're referring to.

> the problem is we don't know the N for each packed batch and N changes per batch. For example, if N=331 for a given packed batch and batch_size=128, without packing it is quite a headache to replicate, (although I agree is theoretically possible).

Of course - sorry if it was not clear, but I meant that N changes across batches. Note that, in our training setup, batch size has a limited impact assuming you end up ultimately training on the same number of examples.

> I looked at the MTF code but didn't notice where the loading of the Adafactor parameters is but I'll keep looking.

It's simply part of the checkpoint loading. The checkpoint has the accumulator variables, so when you load the pre-trained checkpoint it loads the accumulator variables into their associated nodes in the TF graph. There is no special code to do this.

> If you can somehow quantify the generalization difference you saw (even raw estimation) that would be useful so we could know what to expect.

We did not experiment much with this.

**Comment by saareliad** [comment]
Thanks @craffel 
The aforementioned colab
https://colab.research.google.com/github/patil-suraj/exploring-T5/blob/master/T5_on_TPU.ipynb
Finetuning with it produces worse results than C4 checkpoint.

I think that MTF packing affects the result in some way which we can't easily quantify.
The naive way of doing packing is fixing N per batch (row per example) and trimming by the longest seq length.
This, of course, is not efficient in TPUs (Systolic Arrays) so I understand why you did it.

But to fully replicate MTF experiments without packing (or with naive packing):
* one would have to pre-check how many examples each step() trained on, (check N for every batch), then accumulate N examples.
* This may not be the end: unless something in MTF implementation prevents it, for a batch with "long" examples N will be small (e.g N=8), and therefore each example can have a higher impact on the step  (since the loss is usually averaged over batch dimension) compared to a batch with N=100 short examples.

I tried to look and verify how packing in MTF is implemented end-to-end and check the relation with attention cross-entropy and so on, with goal in mind to implement similar things in PyTorch
(e.g follow from here
https://github.com/tensorflow/mesh/blob/6a812c8bb847e081e976533ed497c7c5016bb1ec/mesh_tensorflow/transformer/dataset.py
to here
https://github.com/tensorflow/mesh/blob/6f5e3f10b5fe2bbd613bbe11b18a63d52f2749f4/mesh_tensorflow/transformer/utils.py#L598
all the way to internal operations 
and how they read position/segmentation/ids from the `Context` object)
But I kind of got lost in the details for one day

**Comment by craffel** [comment]
I am completely on the same page as you that replicating packing with padding (in terms of having the same # of examples in a packed vs. padded batch) is impractical if not impossible. However, in all of our experiments the most important factor was how many tokens the model was trained on, regardless of how many tokens (or examples) were in each batch. In other words, if you train on X total tokens with packing or you train on X total tokens with padding, I would expect your results to be very similar.

BTW, one other outstanding issue I was reminded of between HF and MTF T5 is being discussed here: https://github.com/huggingface/transformers/issues/5543
